circle/src/point.rs

```diff
@@ -0,0 +1,165 @@
+    _private: (),
+}
+
+impl<F: Field> Point<F> {
+    #[inline]
+    pub fn new(x: F, y: F) -> Point<F> {
+        debug_assert_eq!(x.square() + y.square(), F::one());
+        Point { x, y, _private: () }
+    }
+    pub fn zero() -> Self {
+        Self::new(F::one(), F::zero())
+    }
+
+    /// Circle STARKs, Section 3, Lemma 1: (page 4 of the first revision PDF)
+    /// ```ignore
+    /// (x, y) = ((1-t^2)/(1+t^2), 2t/(1+t^2))
+    /// ```
+    /// Panics if t^2 = -1, corresponding to either of the points at infinity
+    /// (on the projective *circle*) (1 : ±i : 0)
+    pub fn from_projective_line(t: F) -> Self {
+        let t2 = t.square();
+        let inv_denom = (F::one() + t2).try_inverse().expect("t^2 = -1");
+        Self::new((F::one() - t2) * inv_denom, t.double() * inv_denom)
+    }
+
+    /// Circle STARKs, Section 3, Lemma 1: (page 4 of the first revision PDF)
+    /// ```ignore
+    /// t = y / (x + 1)
+    /// ```
+    /// Returns None if self.x = -1, corresponding to Inf on the projective line
+    ///
+    /// This is also used as a selector polynomial, with a simple zero at (1,0)
+    /// and a simple pole at (-1,0), which in the paper is called v_0
+    /// Circle STARKs, Section 5.1, Lemma 11 (page 21 of the first revision PDF)
+    pub fn to_projective_line(self) -> Option<F> {
+        self.y.try_div(self.x + F::one())
+    }
+
+    /// The "squaring map", or doubling in additive notation, denoted π(x,y)
+    /// Circle STARKs, Section 3.1, Equation 1: (page 5 of the first revision PDF)
+    pub fn double(self) -> Self {
+        Self::new(
+            self.x.square().double() - F::one(),
+            self.x.double() * self.y,
+        )
+    }
+
+    /// Evaluate the vanishing polynomial for the standard position coset of size 2^log_n
+    /// at this point
+    /// Circle STARKs, Section 3.3, Equation 8 (page 10 of the first revision PDF)
+    pub fn v_n(mut self, log_n: usize) -> F {
+        for _ in 0..(log_n - 1) {
+            self.x = self.x.square().double() - F::one();
+        }
+        self.x
+    }
+
+    /// Evaluate the formal derivative of `v_n`
+    /// Circle STARKs, Section 5.1, Remark 15 (page 21 of the first revision PDF)
+    pub fn v_n_prime(self, log_n: usize) -> F {
+        F::two().exp_u64((2 * (log_n - 1)) as u64) * (1..log_n).map(|i| self.v_n(i)).product()
+    }
+
+    /// Evaluate the single-point vanishing function v_p(x). Used for DEEP quotient.
+    /// Simple zero at p, simple pole at +-infinity.
+    /// Circle STARKs, Section 3.3, Equation 11 (page 11 of the first edition PDF).
+    pub fn v_p<EF: ExtensionField<F>>(self, p: Point<EF>) -> (EF, EF) {
+        let x_minus_p = -p + self;
+        (EF::one() - x_minus_p.x, -x_minus_p.y)
+    }
+
+    /// The concrete value of the selector s_P = v_n / (v_0 . T_p⁻¹) at P, used for normalization to 1.
+    /// Circle STARKs, Section 5.1, Remark 16 (page 22 of the first revision PDF)
+    pub fn s_p(self, log_n: usize) -> F {
```

Comment:
nitpick: a bit misleading name, I would propose something that clarifies that it is `s_P` at `P`.